src/main/java/com/yelp/nrtsearch/server/luceneserver/field/properties/Keyable.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2020 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.field.properties;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.Term;
+
+/**
+ * Trait interface for {@link com.yelp.nrtsearch.server.luceneserver.field.FieldDef} types that can
+ * be used for keying Documents.
+ *
+ * <p>There are two purposes for Keyable fields: 1. Only one document can exists with the same key
+ * value 2. An AddDocument request with an existing key value will update the document
+ */
+public interface Keyable {
+
+  /**
+   * Get the name of the keyable field This will be used to generate the Term for updating the
+   * document
+   *
+   * @return name of the keyable field
+   */
+  String getName();
+
+  /**
+   * One FieldDef has to be registered with key=true in order to be used for updating the documents.
+   *
+   * @return whether the field is registered with key=true
+   */
+  boolean isKey();
```

Comment:
I'm not sure that these methods are needed on the interface. getName can be provided by the FieldDef.

isKey may not need to be exposed publicly. The index key can be identified from the IndexState. If we did want to expose it, it might be better as a property of FieldDef/IndexableFieldDef.

Reply:
I was able to get rid of `isKey()`. It might be okay to leave the `getName()` since it's anyway going to be retrieved from the FieldDef and would avoid having to get it from FieldDef and pass it around to Keyable instances.